chargebee/src/main/java/com/chargebee/android/billingservice/BillingClientManager.kt

```diff
@@ -253,6 +255,102 @@ class BillingClientManager(context: Context) : PurchasesUpdatedListener {
 
     }
 
+    internal fun changeProduct(
+        changeProductParams: ChangeProductParams,
+        purchaseCallBack: CBCallback.PurchaseCallback<String>
+    ) {
+        this.purchaseCallBack = purchaseCallBack
+        onConnected({ status ->
+            if (status) {
+                changeProduct(changeProductParams)
+            } else
+                purchaseCallBack.onError(
+                    connectionError
+                )
+        }, { error ->
+            purchaseCallBack.onError(error)
+        })
+
+    }
+
+    private fun changeProduct(changeProductParams: ChangeProductParams) {
+        this.changeProductParams = changeProductParams
+        this.purchaseProductParams = changeProductParams.purchaseProductParams
+        val offerToken = changeProductParams.purchaseProductParams.offerToken
+        val oldProductId = changeProductParams.oldProductId
+        val prorationMode = changeProductParams.prorationMode ?: BillingFlowParams.ProrationMode.IMMEDIATE_WITH_TIME_PRORATION
+
+        val queryProductDetails = arrayListOf(QueryProductDetailsParams.Product.newBuilder()
+            .setProductId(this.purchaseProductParams.product.id)
+            .setProductType(this.purchaseProductParams.product.type.value)
+            .build())
+
+        val productDetailsParams = QueryProductDetailsParams.newBuilder().setProductList(queryProductDetails).build()
+
+        billingClient?.queryProductDetailsAsync(
+            productDetailsParams
+        ) { billingResult, productsDetail ->
+            if (billingResult.responseCode == OK && productsDetail != null) {
+                val productDetailsBuilder = BillingFlowParams.ProductDetailsParams.newBuilder()
+                    .setProductDetails(productsDetail.first())
+                offerToken?.let { productDetailsBuilder.setOfferToken(it) }
+                val productDetailsParamsList =
+                    listOf(productDetailsBuilder.build())
+
+                queryAllSubsPurchaseHistory(ProductType.SUBS.value) { subscriptionHistory ->
+                    val oldPurchaseToken: String = getOldPurchaseToken(subscriptionHistory, oldProductId)
+
+                    val billingFlowParams =
+                        BillingFlowParams.newBuilder()
+                            .setProductDetailsParamsList(productDetailsParamsList)
+                            .setSubscriptionUpdateParams(
+                                BillingFlowParams.SubscriptionUpdateParams.newBuilder()
+                                    .setOldPurchaseToken(oldPurchaseToken)
```

Comment:
What would be the response if the `oldPurchaseToken` is empty/null?

Reply:
It's loading infinitely. Handled it now with an error message.